integration/tbot_test.go

```diff
@@ -0,0 +1,224 @@
+/*
+ *
+ * Copyright 2022 Gravitational, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * /
+ */
+
+package integration
+
+import (
+	"context"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	"github.com/stretchr/testify/require"
+
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/auth"
+	"github.com/gravitational/teleport/lib/config"
+	"github.com/gravitational/teleport/lib/service"
+	"github.com/gravitational/teleport/lib/tbot"
+	"github.com/gravitational/teleport/lib/tbot/testhelpers"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+func rotate(
+	ctx context.Context, t *testing.T, log logrus.FieldLogger, svc func() *service.TeleportProcess, phase string, reload chan struct{},
+) {
+	t.Helper()
+
+	log.Infof("Triggering rotation: %s", phase)
+	err := svc().GetAuthServer().RotateCertAuthority(ctx, auth.RotateRequest{
+		// only rotate Host CA as to avoid race condition serverside when
+		// multiple CAs are rotated at once and the database closes off.
+		Type:        types.HostCA,
+		Mode:        types.RotationModeManual,
+		TargetPhase: phase,
+	})
+	if err != nil {
+		log.WithError(err).Infof("Error occurred during triggering rotation: %s", phase)
+	}
+	require.NoError(t, err)
+	log.Infof("Triggered rotation: %s", phase)
+
+	if phase == types.RotationPhaseInit {
+		err = waitForProcessEvent(svc(), service.TeleportPhaseChangeEvent, 30*time.Second)
+		require.NoError(t, err)
+	} else {
+		err = waitForProcessEvent(svc(), service.TeleportReloadEvent, 30*time.Second)
```

Comment:
isn't it a race if the service already reloaded and the return from `svc()` is the new one?